README.md
# react-d3-heatmap

![Build](https://travis-ci.org/Thebarda/react-d3-heatmap.svg?branch=master)

A simple React heat map to visualize your data between two dates using d3.js

This heat map will display squares inside SVG. A square represent a day.

Now, it supports Typescript 🎉

# Demo

<a href="https://wnk771y0m7.codesandbox.io" target="_blank">Here is a preview</a>
Or
<a href="https://codesandbox.io/s/wnk771y0m7" target="_blank">View with the whole code</a>

# Installation

    npm i react-d3-heatmap

or

    yarn add react-d3-heatmap

# Usage

react-d3-heatmap is very easy to use. 

Here is a basic usage :
```javascript
<HeatMapDate
    startDate={startDate}
    endDate={endDate}
    data={data}
    colors={colors}
    displayLegend
/>
```

# Documentation

You can change these props :

| Prop | Definition | Type | Is required ? | Default |
|------|-------------|------|---------------|---------|
|data|Your data to visualize. <br /> It must be an array of objects that contain a date and a number|`Array[{date: Date, count: Number}]`|`true`|
|startDate|Your heat map will begin at this date| `Date`|`true`|
|endDate|Your heat map will end at this date|`Date`|`true`|
|colors|Let's colorize this heat map <br /> It must be an array of objects that a number and a color like `"#232323"`|`Array[{count: Number, color: String, text: string (optional)}]`|`true`|
|defaultColor|A string that contain a default color|`String`|`false`|`"#cdcdcd"`
|textDefaultColor|A tooltip's text displayed in legend for default color|`String`|`false`|`"0"`
|rectWidth|Set width and height for each square|`Number`|`false`|`10`
|marginRight|Space between two squares on x axis|`Number`|`false`|`4`
|marginBottom|Space between two squares on y axis|`Number`|`false`|`4`
|displayLegend|Display or not a legend|`Boolean`|`false`|`true`
|backgroundColor|Custom background color (ex: `"#232323"`)|`String`|`false`|`"#fff"`
|textColor|Custom text color (ex: `"#00ff00"`)|`String`|`false`|`"#000"`
|radius|Apply a radius on each square|`Number`|`false`|`0`
|classnames|Apply your own CSS class|`String`|`false`|`""`
|displayYear|Display on each month (ex: `17/Avr`)|`Boolean`|`false`|`false`
|onClick|Execute a callback function when clicking on a square. Params: `(data, index)`|`Function`|`false`|`{}`
|onMouseEnter|Execute a callback function when mouse entering in a square. Params: `(data, index)`|`Function`|`false`|`{}`
|onMouseLeave|Execute a callback function when mouse leaving a square. Params: `(data, index)`|`Function`|`false`|`{}`
|shouldStartMonday|Usually heatmap starts on Sunday but you can start it on Monday|`Boolean`|`false`|`false`
|monthSpace|Change space between month blocks|`Number`|`false`|`0`
|rangeDisplayData|Display data between two dates. Tip: You can create an array of one or two dates <br > Example: `[new Date("2018-10-15T17:41:27"), new Date("2019-02-21T23:01:50")]`, `[new Date("2019-02-21T23:01:50")]`, `[null, new Date("2019-02-21T23:01:50")]` |`Array[Date]`|`false`|`[]`
|fadeAnimation|Control fade in animation|`Object{animate: Boolean, duration: Number (in second)}`|`false`|`{ animate: true, duration: 0.4 }`

# Author

Tom Darneix
# License

MIT


CHANGELOG.md
# 1.1.5

New Feature : 
  - fadeAnimation : Control fade in animation. 

Other :
  - Apply patches for High severities

# 1.1.4

New Feature : 
  - rangeDisplaData : Display data between two dates. 

# 1.1.3

New Feature : 
  - monthSpace : You can change space between months

Enhancement : 
  - Improve months display

Other : 
  - Refactor code

# 1.1.1

Security : 
  - Resolve vulnerability alert from `webpack-dev-server`. See vulnerability [Here](https://www.npmjs.com/advisories/725)

# 1.1.0

Enhancement:
  - Support typescript 🎉

Important : Transition has been removed

# 1.0.9

New Feature:
  - You can choose to start your heatmap Monday or Sunday. Default is Sunday

Security alert:
  - Fix a dependency alert (dep affected : merge < 1.2.1)

# 1.0.8

New Features:
  - Add a function that convert an array of dates to an array of data (`Array[{date: Date, count: Number}]`)
  - You can custom tooltip's text
  - You can also custom default color tooltip's text

Improvement:
  - Default color is displayed in legend

# 1.0.7

New Features:
  - You can custom backgroundColor, textColor, squares radius, CSS class.
  - You can choose to displayYear
  - You can execute your own function when mouse clicks, enters and/or leaves a square
  - When a square is hovered, it display a stroke on this square

Bug fix:
  - Legend tooltips should works correctly even if datas are updated

# 1.0.6

Bug fix:
  - Tooltips should works correctly even if datas are updated

# 1.0.3

Bug fix

# 1.0.1

Bug fixes:
  - Sometime tooltip does not display (due to a workaround)

# 1.0.0

New Feature:
  - You can add a transition when heat map renders for the first time
  - You can custom background and text color

Improvements:
  - Improve tests

Bug fixes

# 0.0.17-beta

bug fix: 
  - When legend is disabled, we set both width and height to 0 
  - expand svg to the right + improve month display by display month when startDate and endDate have same month and year
  - update README and remove a log print 

# 0.0.15-beta

New Feature: 
  - Heat map to visualize data between two dates


src/components/HeatMapDate.js
import React, { PureComponent } from "react"
import PropTypes from "prop-types"
import * as d3 from "d3"
import d3Tip from "d3-tip"

/**
 * Component that display a heatmap to visualize data through date.
 * Each square is a day.
 */
export default class HeatMapDate extends PureComponent {
	/**
	 * Props type checking
	 */
	static propTypes = {
		// The 'visible' heatmap will start this date
		startDate: PropTypes.instanceOf(Date).isRequired,
		// The 'visible' heatmap will end this date
		endDate: PropTypes.instanceOf(Date).isRequired,
		// The data that fill the heatmap. Must be an Array[{data: Date, count: Number}]
		data: PropTypes.instanceOf(Array).isRequired,
		// Colors that apply a color on squares. Must be an Array[{count: Number, color: String}]
		colors: PropTypes.instanceOf(Array).isRequired,
		// Apply a default color for dates whose count are too low to apply a color from 'colors'
		defaultColor: PropTypes.string,
		// Custom text for default color in tooltip legend
		textDefaultColor: PropTypes.string,
		// Define a width and height for square
		rectWidth: PropTypes.number,
		// Define a margin between two squares on x axis
		marginRight: PropTypes.number,
		// Define a margin between two squares on y axis
		marginBottom: PropTypes.number,
		// Display a legend or not
		displayLegend: PropTypes.bool,
		// Apply a transition when heatmap is rendering for the first time
		transition: PropTypes.number,
		// Apply a background color
		backgroundColor: PropTypes.string,
		// Apply a text color (unavailable on tooltip)
		textColor: PropTypes.string,
		// Apply a radius on square
		radius: PropTypes.number,
		// Class attributes,
		classnames: PropTypes.string,
		// Display year near each month
		displayYear: PropTypes.bool,
		// Handle onClick in container callback
		onClick: PropTypes.func,
		// Handle onMouseEnter on each square callback
		onMouseEnter: PropTypes.func,
		// Handle onMouseLeave on each square callback
		onMouseLeave: PropTypes.func,
		// Change week-day to start (Sunday or Monday)
		shouldStartMonday: PropTypes.bool,
	}

	/**
	 * Set a default value to unrequired props
	 */
	static defaultProps = {
		marginRight: 4,
		marginBottom: 4,
		displayLegend: true,
		rectWidth: 10,
		defaultColor: "#cdcdcd",
		transition: -1,
		backgroundColor: "#fff",
		textColor: "#000",
		radius: 0,
		classnames: "",
		onClick: () => {},
		onMouseLeave: () => {},
		onMouseEnter: () => {},
		shouldStartMonday: false,
	}

	constructor(props) {
		super(props)
		this.ID = Math.random()
			.toString(36)
			.replace(/[^a-z]+/g, "")
			.substr(0, 32)
		this.IDLegend = Math.random()
			.toString(36)
			.replace(/[^a-z]+/g, "")
			.substr(0, 32)
	}

	state = {
		svgElem: undefined,
		svgLegend: undefined,
		firstRender: true,
	}

	componentDidUpdate() {
		// I use a setTimeout to prevent a component update that stop the transition.
		setTimeout(() => {
			if (this.props.transition > 0 && this.state.firstRender && this.state.svgElem) {
				this.setState({ firstRender: false })
			}
		}, this.props.transition)
	}

	render() {
		const {
			startDate,
			endDate,
			data,
			colors,
			defaultColor,
			rectWidth,
			marginRight,
			marginBottom,
			displayLegend,
			transition,
			backgroundColor,
			textColor,
			radius,
			classnames,
			displayYear,
			onClick,
			onMouseEnter,
			onMouseLeave,
			textDefaultColor,
			shouldStartMonday,
		} = this.props
		const { svgElem, svgLegend, firstRender } = this.state
		// Array of months for x axis
		const monthsName = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "July", "Aug", "Sep", "Oct", "Nov", "Dec"]
		// Array of days for y axis
		const daysName = !shouldStartMonday ? ["Sun", "Tue", "Thu", "Sat"] : ["Mon", "Wed", "Fri", "Sun"]
		const dataset = []
		let t = null
		if (transition > 0 && firstRender) {
			t = d3.transition().duration(transition)
		}

		// This is a possible workaround about tooltips that do not want to hide when data changes
		// See https://github.com/Caged/d3-tip/issues/133
		d3.select(".d3-tip." + this.ID).remove()
		d3.select(".d3-tip." + this.IDLegend).remove()

		const svg = d3.select(svgElem)
		// We remove all elemnts (rect + text) to properly update the svg
		svg.selectAll("*").remove()
		const tmpBufferDate = new Date(startDate)
		const startDateYesterday = new Date(startDate)
		// When want to display month on first column if difference between
		// startDate and endDate less than 1 month
		const noMonthName =
			(startDate.getMonth() === endDate.getMonth() && startDate.getFullYear() === endDate.getFullYear()) ||
			(startDate.getMonth() == 11 &&
				endDate.getMonth() === 0 &&
				endDate.getFullYear() - startDate.getFullYear() === 1)
		startDateYesterday.setDate(startDateYesterday.getDate() - 1)
		// We set bufferDate to the previous Sunday (or Monday following 'shouldStartMonday' prop) of startDate.
		tmpBufferDate.setDate(tmpBufferDate.getDate() - startDateYesterday.getDay())
		if (!shouldStartMonday) {
			tmpBufferDate.setDate(tmpBufferDate.getDate() - 1)
		}
		const bufferDate = new Date(tmpBufferDate)
		bufferDate.setHours(0, 0, 0, 0)
		// Number of day from bufferDate to endDate
		const nbDayDiff = (endDate.getTime() - bufferDate.getTime()) / 1000 / 60 / 60 / 24
		const svgWidth = (rectWidth + marginRight) * (nbDayDiff / 7) + 70
		svg.attr("width", svgWidth).attr("height", (rectWidth + marginBottom) * 7 + 50)

		if (noMonthName) {
			const prefix = displayYear
				? new Date(bufferDate)
						.getFullYear()
						.toString()
						.substring(2, 4) + "/"
				: ""
			const prefixWidth = displayYear ? 25 : 0
			svg.append("text")
				.text(prefix + monthsName[startDate.getMonth()])
				.attr("x", () => {
					return Math.floor(0 / 7) * (rectWidth + marginRight) + 32 - prefixWidth
				})
				.attr("y", 18)
				.attr("font-size", 18)
				.attr("fill", textColor)
		}

		for (let i = 0; i < nbDayDiff; i++) {
			const prefix = displayYear
				? new Date(bufferDate)
						.getFullYear()
						.toString()
						.substring(2, 4) + "/"
				: ""
			const prefixWidth = displayYear ? 25 : 0
			if (i == 0 || i === 2 || i === 4 || i === 6) {
				// Display day name as y axis
				svg.append("text")
					.text(daysName[i / 2])
					.attr("y", (i % 7) * (rectWidth + marginBottom) + rectWidth / 6 + 32)
					.attr("x", 0)
					.attr("fill", textColor)
			}
			// Find the first data that match with current bufferDate
			const objMatch = data.find(obj => {
				const dateTmp = new Date(obj.date)
				dateTmp.setHours(0, 0, 0, 0)
				bufferDate.setHours(0, 0, 0, 0)
				return dateTmp.getTime() === bufferDate.getTime()
			})
			// If bufferDate < (startDate - 1 day) we set the square color like background to make that 'invisible'
			let finalColor = backgroundColor
			let maxCount = null
			// If there is no match we set the default color
			if (objMatch === undefined && bufferDate.getTime() >= startDateYesterday.getTime()) {
				finalColor = defaultColor
			} else if (bufferDate.getTime() >= startDateYesterday.getTime()) {
				finalColor = colors.filter(c => c.count <= objMatch.count)
				if (finalColor.length === 0) {
					finalColor = defaultColor
				} else {
					finalColor = finalColor[finalColor.length - 1].color
				}
			}
			const today = new Date(bufferDate.getTime())
			// Finally, we push it to an Array that will be used by d3
			dataset.push({ date: today, count: objMatch ? objMatch.count : maxCount || 0, color: finalColor, i })

			if (bufferDate.getDate() === 1 && !noMonthName) {
				// Display month name
				svg.append("text")
					.text(prefix + monthsName[bufferDate.getMonth()])
					.attr("x", () => {
						return Math.floor(i / 7) * (rectWidth + marginRight) + 32 - prefixWidth
					})
					.attr("y", 18)
					.attr("font-size", 18)
					.attr("fill", textColor)
			}
			bufferDate.setDate(bufferDate.getDate() + 1)
		}

		if (dataset.length > 0) {
			// I added an ID the tooltip because it's a workaround to prevent the tooltip won't hide when the component is updating
			const tip = d3Tip()
				.attr("class", "d3-tip " + this.ID)
				.offset([-8, 0])
				.html(d => {
					if (d.color !== backgroundColor) {
						return (
							"<div style={{ fontSize: '15' }}>" +
							d.date.getFullYear() +
							"/" +
							(d.date.getMonth() + 1) +
							"/" +
							d.date.getDate() +
							" : " +
							d.count +
							"</div>"
						)
					} else return null
				})
			svg.call(tip)
			// Display all data squares
			const rects = svg
				.selectAll("rect")
				.data(dataset)
				.enter()
				.append("rect")
				.attr("fill-opacity", t !== null ? 0 : 1)
				.attr("width", rectWidth)
				.attr("height", rectWidth)
				.attr("class", "dayRect")
				.attr("x", d => {
					return Math.floor(d.i / 7) * (rectWidth + marginRight) + 40
				})
				.attr("y", d => {
					return (d.i % 7) * (rectWidth + marginBottom) + 24
				})
				.attr("fill", d => d.color)
				.attr("rx", radius)
				.attr("ry", radius)
				.on("mouseover", function(d, i) {
					if (d.color !== backgroundColor) {
						tip.show(d, this)
						d3.select(this).attr("stroke", "black")
					}
					onMouseEnter(d, i)
				})
				.on("mouseout", (d, i) => {
					tip.hide(d, this)
					d3.selectAll("rect").attr("stroke", "none")
					onMouseLeave(d, i)
				})
				.on("click", (d, i) => {
					onClick(d, i)
				})

			if (t !== null) rects.transition(t).attr("fill-opacity", 1)
		}

		let legendWidth = 0

		if (displayLegend) {
			const svgLegendD3 = d3.select(svgLegend)
			svgLegendD3.selectAll("*").remove()
			legendWidth = (rectWidth + marginRight) * colors.length + 90 + 50
			svgLegendD3.attr("width", legendWidth).attr("height", 30)
			svgLegendD3
				.append("text")
				.text("Legend :")
				.attr("x", 0)
				.attr("y", 20)
				.attr("font-size", 18)
				.attr("fill", textColor)

			const tip = d3Tip()
				.attr("class", "d3-tip " + this.IDLegend)
				.offset([-8, 0])
				.html(d => {
					const displ = d.text ? d.text : d.count.toString()
					return "<div style={{ fontSize: '15' }}>" + displ + "</div>"
				})
			svgLegendD3.call(tip)
			svgLegendD3
				.selectAll("rect")
				.data(
					[
						{
							color: defaultColor,
							count: 0,
							text: textDefaultColor ? textDefaultColor : "0",
						},
					].concat(colors)
				)
				.enter()
				.append("rect")
				.attr("width", rectWidth)
				.attr("height", rectWidth)
				.attr("x", (d, i) => (rectWidth + marginRight) * i + 80)
				.attr("y", 15 - rectWidth / 2)
				.attr("rx", radius)
				.attr("ry", radius)
				.attr("fill", d => d.color)
				.on("mouseover", function(d) {
					if (d.color !== backgroundColor) {
						tip.show(d, this)
						d3.select(this).attr("stroke", textColor)
					}
				})
				.on("mouseout", d => {
					tip.hide(d, this)
					d3.selectAll("rect").attr("stroke", "none")
				})
		} else {
			const svgLegendD3 = d3.select(svgLegend)
			svgLegendD3.attr("width", 0).attr("height", 0)
		}

		return (
			<div
				className={classnames}
				style={{
					width: legendWidth > svgWidth ? legendWidth : svgWidth + "px",
					height: "auto",
					backgroundColor: backgroundColor,
				}}
				id="react-d3-heatMap">
				<svg
					style={{ display: "block" }}
					ref={elem => {
						if (!this.state.svgElem) this.setState({ svgElem: elem })
					}}
				/>
				<svg
					ref={elem => {
						if (!this.state.svgLegend) this.setState({ svgLegend: elem })
					}}
				/>
			</div>
		)
	}
}
